### ISSUE TITLE
Support "native" podman as backend

### ISSUE THREAD

**Issue by AkiraNorthstar** [woodpecker-ci/woodpecker#85]
Hello Laszlo!
Is it possible to integrate podman in woodpecker?

Podman does not run as a daemon (like docker : /var/run/docker.sock) but is fully compatible with the docker command line.

Another point is that podman can also mount secrets via [mount.conf](https://github.com/containers/libpod/blob/master/install.md#mountsconf), then you maybe have also solved [Global Secrets](https://github.com/laszlocph/woodpecker/issues/7).

Also cgroups v2 is native supported by podman.

Further information about podman:
- https://podman.io/
- https://github.com/containers/libpod

**Comment by laszlocph** [comment]
Technically it is a matter of implementing this interface: https://github.com/laszlocph/woodpecker/blob/master/cncd/pipeline/pipeline/backend/backend.go and this is how it is implemented for Docker: https://github.com/laszlocph/woodpecker/blob/master/cncd/pipeline/pipeline/backend/docker/docker.go

So technically it is probably possible. As for should the project focus on it, is a different question.

In my surroundings I don't see many companies want to differentiate on what container engines they are running, and just default on Docker. Even though the drawbacks of Docker's architecture is known to some of them, adopting alternatives have not reached a high level. Or they don't talk about it, or I don't listen :) Also these companies are not Fedora/RedHat/Centos shops, so I have limited visibility on the adoption of Podman. But I do see the benefits of Podman over Docker. I haven't used it, but I might try.

All in all:
- technically it's possible
- I don't plan to immediately implement it
- I accept pull requests even in early forms, and happy to team up in full delivery
- I need more data on the adoption of Podman.

**Comment by mscherer** [comment]
So I tried to use podman with woodpecker. It failed in several interesting ways. I used podman on f34, either the package (quite recent 3.3.0), or the latest git devel version (self compiled). 

It failed first because podman didn't pull the plugins/git image, and I do not understand why as it work if I do it manually. Once the plugin is in the local registry, the build proceed.

Then it failed because the workingDir directory is not created automatically. I see the volume creation, and the code should create it on disk in /var/lib/something, but it doesn't seems to mount it correctly. Again, it seems to be dependent on the backend, i tried with crun and runc, no luck.  My plan was to debug that, but I haven't yet looked more. 

However, I also looked at what it would requires to write a backend. Podman has a set of [bindings](https://github.com/containers/podman/tree/main/pkg/bindings), so this shouldn't have been too hard for a go beginner like me. However, since woodpecker has switched to vendoring in 75513575be , my understanding is that we also have to vendor podman, which in turn pull a rather large number of dependencies, some who requires C headers (btrfs, devicemapper). I do think this would put a undue burden on woodpecker CI and deployment.

So for now, the solution for using podman is either find why woodpecker work with docker but not the docker API exposed by podman (likely a issue podman side), or add a backend in a way that do not bloat the build (again, likely a issue on podman binding, or maybe I do not it wrong ).

**Comment by anbraten** [comment]
That sounds quite interesting. I never used podman, but as I am totally interested in a Kubernetes agent, it is probably a good discussion about how backends for agent should be handled in the long term. Currently I see two options for it:
- separate agents per backend
- one agent which includes every supported backend

Meanwhile if you need help debugging woodpecker feel free to write me on [Discord](https://discord.gg/fcMQqSMXJy).

**Comment by mfulz** [comment]
> That sounds quite interesting. I never used podman, but as I am totally interested in a Kubernetes agent, it is probably a good discussion about how backends for agent should be handled in the long term. Currently I see two options for it:
> 
> * separate agents per backend
> * one agent which includes every supported backend
> 
> Meanwhile if you need help debugging woodpecker feel free to write me on [Discord](https://discord.gg/fcMQqSMXJy).

I've just used a --use-podman boolean flag for the agent for now.
But I think in the actual Implementation it would make sense to use one agent with multiple backends and implementing some flag like --backend with a simple switch...case inside the agent.

Most of the code would be either just duplicated for every agent or it would need a bigger rewrite I guess.

**Comment by mscherer** [comment]
So I just tried again with podman and it now work.... almost.

I was able to use on a self hosted gitea the following config:
```
clone:
  git:
    image: docker.io/a6543/test_git_plugin:latest
pipeline:
  prepare-build:
    image: quay.io/fedora/fedora:latest
    commands:
      - sudo dnf install -y zola
      - zola build --drafts
```

The server run on Fedora 34 with podman-3.3.1-1.fc34.x86_64, with woodpecker 34cfabb5 
So the issue of not downloading the image is however still here, but there is a workaround unlike the working dir one.

**Comment by mscherer** [comment]
Ok so it worked because I patched podman myself, and forgot about it. Upgrading to 3.4.0 resulted in the same workingDir issue

**Comment by mscherer** [comment]
Status update, I managed to get podman working (for real this time), and the patch was merged upstream, and will be in the next release. Now the only last issue is one around registry. It seems for some reason, the docker compat API of podman will not download remote images automatically, but if you download manually before (using podman pull on the agent), it work without problem, so this one can be worked around more easily.

**Comment by 6543** [comment]
@mscherer nice to hear - so do we need #305 or do work podman with compatiblity layer with podman >= v3.4.3 (upcomming release) ?

**Comment by mscherer** [comment]
I think it will work with newer podman, but I would appreciate someone to do a test, since I already managed to get my test wrong.
I also wrote https://github.com/mscherer/podman/commit/bf4a6b99b9cccb3e00ef498e4a49e7b6e56d4b2c to fix the 2nd problem I faced, but I need to research a bit more the code of docker to make sure I replicated the behavior correctly.

**Comment by 6543** [comment]
looking forward do see a pull for the 2nd issue :)

I'll think we should still add podman as "native" option too - but what ever get released first will do it for the majority i think

**Comment by mscherer** [comment]
Here is the PR for the last problem I encountered: 
https://github.com/containers/podman/pull/12315

**Comment by mscherer** [comment]
So the PR was wrong, but this is being worked on https://github.com/containers/podman/pull/12318 and https://github.com/containers/podman/pull/12317 .

In the mean time, I also found another incompatibility on https://github.com/containers/podman/issues/12320 (which might be trickier to solve)

**Comment by mscherer** [comment]
So, after discussion with the podman devs, we reached a consensus that the compatibility issue is tricky to solve, but it should be fixed. However, my fix is not the right one, and as I do not think there is no easy fix, it might take some time before it get done.

In the mean time, there is a few way that can be explored on woodpecker side:

- convince moby devs to not shorten the name when a full name is given. Eg, change https://github.com/moby/moby/blob/master/client/image_pull.go#L22-L29 
- do not use docker.io for the images request. I stumbled on the bug because I used the custom image made by @6543 on https://github.com/woodpecker-ci/woodpecker/issues/303#issuecomment-932629969 , but if the image was on quay.io, the bug would not be have been triggered. Hence my question regarding self hosting the plugins image or having a vhost dedicated to it in #303
- instruct people wishing to use podman to change the configuration to only search docker.io ( registries.search in /etc/containers/registries.conf ). On Fedora/Centos/RHEL, there is at least Quay and Docker.io, so podman will not allow short name without asking, but maybe it work on Debian without any specific configuration
- ask those users to add shortname alias for the plugin they use. Bonus, this allow to say "git" and do the right thing
- instruct people to download the image manually before using it
- add a  little http proxy between podman and woodpecker-agent to change the parameter

(to be clear, the last one is not a serious proposal)

**Comment by 9p4** [comment]
Ok, it seems like using Podman rootless works just fine for me. What I did was manually link `docker.sock` to the user's podman socket and use a version of Podman>=3.4.3 (because of a bug that would not create the paths in storage).

**Comment by mscherer** [comment]
Podman 4.0 was released yesterday, (announce is not yet published, and will likely not be until next week), and it should now fix the short name issue I faced, so that would be working out of the box. I will do a quick test later.

**Comment by 9p4** [comment]
With PR https://github.com/woodpecker-ci/woodpecker/issues/757, setting the `DOCKER_SOCK` variable to a Podman socket works now (so no more linking required).

**Comment by 6543** [comment]
docker compatibility mode works now ... -> #901 - so renamed the issue

**Comment by NewRedsquare** [comment]
Can someone send an example ? i can't get it working using both commands : 
- `podman container run --privileged --rm --tty -v /run/user/1000/podman/podman.sock:/var/user/1000/podman/podman.sock -e WOODPECKER_SERVER=IP -e WOODPECKER_AGENT_SECRET=xxxxxxxxxx -e WOODPECKER_BACKEND=docker -e DOCKER_HOST=unix:///run/user/1000/podman/podman.sock --network=host docker.io/woodpeckerci/woodpecker-agent:latest`

```
{"level":"error","error":"Cannot connect to the Docker daemon at unix:///run/user/1000/podman/podman.sock. Is the docker daemon running?","time":"2022-06-05T17:31:42Z","message":"could not kill container '0_333181564417296953_clone'"}
{"level":"error","error":"Cannot connect to the Docker daemon at unix:///run/user/1000/podman/podman.sock. Is the docker daemon running?","time":"2022-06-05T17:31:42Z","message":"could not remove container '0_333181564417296953_clone'"}
{"level":"error","error":"Cannot connect to the Docker daemon at unix:///run/user/1000/podman/podman.sock. Is the docker daemon running?","time":"2022-06-05T17:31:42Z","message":"could not kill container '0_333181564417296953_stage_0'"}
{"level":"error","error":"Cannot connect to the Docker daemon at unix:///run/user/1000/podman/podman.sock. Is the docker daemon running?","time":"2022-06-05T17:31:42Z","message":"could not remove container '0_333181564417296953_stage_0'"}
{"level":"error","error":"Cannot connect to the Docker daemon at unix:///run/user/1000/podman/podman.sock. Is the docker daemon running?","time":"2022-06-05T17:31:42Z","message":"could not remove volume '0_333181564417296953_default'"}
{"level":"error","error":"Cannot connect to the Docker daemon at unix:///run/user/1000/podman/podman.sock. Is the docker daemon running?","time":"2022-06-05T17:31:42Z","message":"could not remove network '0_333181564417296953_default'"}
{"level":"error","error":"rpc error: code = Unknown desc = Proc finished with exitcode 1, Cannot connect to the Docker daemon at unix:///run/user/1000/podman/podman.sock. Is the docker daemon running?","time":"2022-06-05T17:31:42Z","message":"grpc error: wait(): code: Unknown: rpc error: code = Unknown desc = Proc finished with exitcode 1, Cannot connect to the Docker daemon at unix:///run/user/1000/podman/podman.sock. Is the docker daemon running?"}
{"level":"warn","repo":"romain/cv","build":"23","id":"73","error":"rpc error: code = Unknown desc = Proc finished with exitcode 1, Cannot connect to the Docker daemon at unix:///run/user/1000/podman/podman.sock. Is the docker daemon running?","time":"2022-06-05T17:31:42Z","message":"cancel signal received"}
```

but : `curl -s --unix-socket /run/user/1000/podman/podman.sock http://d/v1.0.0/libpod/info | jq` gives me normal output

- another error when running : `podman container run --privileged --rm --tty -v /run/user/1000/podman/podman.sock:/var/run/docker.sock -e WOODPECKER_SERVER=IP -e WOODPECKER_AGENT_SECRET=xxxxxxxxxx -e WOODPECKER_BACKEND=docker docker.io/woodpeckerci/woodpecker-agent:latest`

```
"level":"error","error":"Error response from daemon: can only kill running containers. xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx is in state created: container state improper","time":"2022-06-05T17:35:10Z","message":"could not kill container '0_5049337694948908527_clone'"}
{"level":"error","error":"Error response from daemon: no container with name or ID \"0_5049337694948908527_step_0\" found: no such container","time":"2022-06-05T17:35:10Z","message":"could not kill container '0_5049337694948908527_stage_0'"}
{"level":"error","error":"Error: No such container: 0_5049337694948908527_step_0","time":"2022-06-05T17:35:10Z","message":"could not remove container '0_5049337694948908527_stage_0'"}
{"level":"error","error":"rpc error: code = Unknown desc = Proc finished with exitcode 1, Error response from daemon: \"slirp4netns\" is not supported: invalid network mode","time":"2022-06-05T17:35:10Z","message":"grpc error: wait(): code: Unknown: rpc error: code = Unknown desc = Proc finished with exitcode 1, Error response from daemon: \"slirp4netns\" is not supported: invalid network mode"}
{"level":"warn","repo":"romain/cv","build":"25","id":"79","error":"rpc error: code = Unknown desc = Proc finished with exitcode 1, Error response from daemon: \"slirp4netns\" is not supported: invalid network mode","time":"2022-06-05T17:35:10Z","message":"cancel signal received"}
```

i'm kinda lost

**Comment by 9p4** [comment]
I think your `DOCKER_HOST` has to be `unix:///var/user/1000/podman/podman.sock`

**Comment by Taywee** [comment]
In case anybody else has issues with the `clone` step never finishing when using podman as a backend, check this containers/podman#19581

Effectively, set the contents of your containers.conf (in my case, it's `~/.config/containers/containers.conf`) to

```toml
[containers]
log_driver="json-file"

[engine]
events_logger="file"
```

I guess it was reporting events to a different location than the podman service expected to find them.  Now I have woodpecker running entirely through podman!  This is important to me, because I need to use CI to run `podman build`, which I couldn't get working in a docker container, no matter what I tried.  I also am really happy to not have to run a rootful container service to run my woodpecker agent.

**Comment by kaylynb** [comment]
It looks like v2 broke woodpecker-agent running in a container via podman.

In v1.0.5 I regularly get these errors when running pipelines but everything still works fine, including buildx, etc:
```
Nov 19 12:35:54 alzirr systemd-woodpecker-agent[5908]: {"level":"error","error":"Error response from daemon: can only kill running containers. d4dba47c6b966c981d03ff1cc89f6679fc7bac975c38e8cf7b252e55
bb8b225d is in state exited: container state improper","time":"2023-11-19T20:35:54Z","message":"could not kill container 'wp_01hfmmqea1258g2a1c93tz73h2_0_stage_0'"}
```

But in v2.0.0 & latest HEAD (237b2257f5633374a3b28babb2a2d5eef8b30b50 at time of this post) it seems to actually fail on this error now:
```
Dec 01 11:51:17 alzirr systemd-woodpecker-agent2[8964]: {"level":"error","error":"rpc error: code = Unknown desc = Step finished with exit code 1, Error response from daemon: can only kill running co
ntainers. 6231ebb6daff67daebbe56144ea658041a087a08d1aa8d9c185901d894b63c35 is in state exited: container state improper","time":"2023-12-01T11:51:17-08:00","message":"grpc error: wait(): code: Unknow
n: rpc error: code = Unknown desc = Step finished with exit code 1, Error response from daemon: can only kill running containers. 6231ebb6daff67daebbe56144ea658041a087a08d1aa8d9c185901d894b63c35 is i
n state exited: container state improper"}
Dec 01 11:51:17 alzirr systemd-woodpecker-agent2[8964]: {"level":"warn","repo":"jaam/skeetcrawl","pipeline":"0","id":"40","error":"rpc error: code = Unknown desc = Step finished with exit code 1, Err
or response from daemon: can only kill running containers. 6231ebb6daff67daebbe56144ea658041a087a08d1aa8d9c185901d894b63c35 is in state exited: container state improper","time":"2023-12-01T11:51:17-0
8:00","message":"cancel signal received"}
```

I didn't have the time to look through the large diff between v1 & v2 but it does appear that this error should actually be caught here: https://github.com/woodpecker-ci/woodpecker/blob/237b2257f5633374a3b28babb2a2d5eef8b30b50/pipeline/backend/docker/docker.go#L361-L366

**Comment by pat-s** [comment]
Both the underlying WP types changed on a major level as well as the podman pkg (v3/v4). This needs a full rewrite, hence I am closing here.

**Comment by djbobyd** [comment]
Hello I've been getting startup errors running the agent on podman since I upgraded to version 3. It was working fine with the 2.x versions. Here is an excerpt from the log 
```
9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
9:10PM INF src/shared/logger/logger.go:102 > log level: trace
9:10PM INF src/cmd/agent/core/config.go:45 > no agent config found at '/etc/woodpecker/agent.conf', start with defaults
9:10PM INF src/cmd/agent/core/agent.go:73 > shutdown of whole agent
9:10PM INF src/cmd/agent/core/agent.go:99 > shutdown healthcheck server ...
9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
9:10PM INF src/shared/logger/logger.go:102 > log level: trace
9:10PM INF src/cmd/agent/core/config.go:45 > no agent config found at '/etc/woodpecker/agent.conf', start with defaults
9:10PM INF src/cmd/agent/core/agent.go:73 > shutdown of whole agent
9:10PM INF src/cmd/agent/core/agent.go:99 > shutdown healthcheck server ...
9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
9:10PM INF src/shared/logger/logger.go:102 > log level: trace
9:10PM INF src/cmd/agent/core/config.go:45 > no agent config found at '/etc/woodpecker/agent.conf', start with defaults
9:10PM INF src/cmd/agent/core/agent.go:73 > shutdown of whole agent
9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
9:10PM INF src/shared/logger/logger.go:102 > log level: trace
9:10PM INF src/cmd/agent/core/config.go:45 > no agent config found at '/etc/woodpecker/agent.conf', start with defaults
9:10PM INF src/cmd/agent/core/agent.go:73 > shutdown of whole agent
9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
9:10PM INF src/cmd/agent/core/agent.go:99 > shutdown healthcheck server ...
```

Anyone has any idea what this error could mean? Thanks!

**Comment by djbobyd** [comment]
> Hello I've been getting startup errors running the agent on podman since I upgraded to version 3. It was working fine with the 2.x versions. Here is an excerpt from the log
> 
> ```
> 9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
> 9:10PM INF src/shared/logger/logger.go:102 > log level: trace
> 9:10PM INF src/cmd/agent/core/config.go:45 > no agent config found at '/etc/woodpecker/agent.conf', start with defaults
> 9:10PM INF src/cmd/agent/core/agent.go:73 > shutdown of whole agent
> 9:10PM INF src/cmd/agent/core/agent.go:99 > shutdown healthcheck server ...
> 9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
> 9:10PM INF src/shared/logger/logger.go:102 > log level: trace
> 9:10PM INF src/cmd/agent/core/config.go:45 > no agent config found at '/etc/woodpecker/agent.conf', start with defaults
> 9:10PM INF src/cmd/agent/core/agent.go:73 > shutdown of whole agent
> 9:10PM INF src/cmd/agent/core/agent.go:99 > shutdown healthcheck server ...
> 9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
> 9:10PM INF src/shared/logger/logger.go:102 > log level: trace
> 9:10PM INF src/cmd/agent/core/config.go:45 > no agent config found at '/etc/woodpecker/agent.conf', start with defaults
> 9:10PM INF src/cmd/agent/core/agent.go:73 > shutdown of whole agent
> 9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
> 9:10PM INF src/shared/logger/logger.go:102 > log level: trace
> 9:10PM INF src/cmd/agent/core/config.go:45 > no agent config found at '/etc/woodpecker/agent.conf', start with defaults
> 9:10PM INF src/cmd/agent/core/agent.go:73 > shutdown of whole agent
> 9:10PM FTL src/cmd/agent/core/run.go:52 > error running agent error="could not create new auth interceptor: rpc error: code = DeadlineExceeded desc = context deadline exceeded"
> 9:10PM INF src/cmd/agent/core/agent.go:99 > shutdown healthcheck server ...
> ```
> 
> Anyone has any idea what this error could mean? Thanks!

I found a workaround. Turns out that somehow the internal connection to port 9000 was not working. I couldn't find out why. I exposed port 9000 to the host and connected the agent to the host IP. It works as expected, although not through the container to container commmunication.

**Comment by gsaslis** [comment]
@djbobyd I am having somewhat similar issues... It would be a big help if you could share how you're running woodpecker-agent and woodpecker-server 

Here 


```
$ cat ~/.config/containers/systemd/woodpecker-server.container
[Container]
Annotation=org.systemd.property.KillMode=none
ContainerName=woodpecker-server
Environment=WOODPECKER_OPEN=True
Environment=WOODPECKER_HOST=https://woodpecker.redacted.domain
<redacted forge env vars>
Environment=WOODPECKER_HOST_URL=https://woodpecker.redacted.domain
Environment=WOODPECKER_LOG_LEVEL=debug
AddHost=woodpecker.redacted.domain:169.254.1.2
Image=woodpeckerci/woodpecker-server:v3-alpine
LogDriver=journald
Network=woodpecker.network
PublishPort=8000:8000
PublishPort=9000:9000
Secret=woodpecker_agent_secret,type=env,target=WOODPECKER_AGENT_SECRET
UserNS=keep-id:uid=1000,gid=1000
Volume=/home/redacted/.local/share/woodpecker:/var/lib/woodpecker
PodmanArgs=--restart on-failure
AutoUpdate=local
[Unit]
After=podman-user-wait-network-online.service

[Install]
WantedBy=default.target
```


and my agent: 

```
$ cat ~/.config/containers/systemd/woodpecker-agent.container
[Container]
Annotation=org.systemd.property.KillMode=none
ContainerName=woodpecker-agent
Environment=WOODPECKER_SERVER=woodpecker-server:9000
AddHost=woodpecker.redacted.domain:169.254.1.2
Exec=agent
Image=woodpeckerci/woodpecker-agent:v3
LogDriver=journald
Network=woodpecker.network
Secret=woodpecker_agent_secret,type=env,target=WOODPECKER_AGENT_SECRET
Volume=/home/radicle/.config/woodpecker:/etc/woodpecker
Volume=/run/user/1001/podman/podman.sock:/var/run/docker.sock
PodmanArgs=--restart on-failure
AutoUpdate=local
[Unit]
Requires=podman.socket
After=podman-user-wait-network-online.service podman.socket

[Install]
WantedBy=default.target
```

The server starts and so does the agent, but when I try to run a pipeline, the agent fails (the pipeline run, it doesn't crash itself) with: 

```
Feb 24 08:21:44 yorgos systemd[1174]: Started woodpecker-agent.service.
Feb 24 08:21:44 yorgos woodpecker-agent[277420]: 1cb96df519751058d1b8bb989c3dab446071f8edc56a4c975acb471d43e00ecf
Feb 24 08:21:44 yorgos woodpecker-agent[277462]: {"level":"info","time":"2025-02-24T08:21:44Z","message":"log level: info"}
Feb 24 08:21:44 yorgos woodpecker-agent[277462]: {"level":"info","time":"2025-02-24T08:21:44Z","message":"starting Woodpecker agent with version '3.1.0' and backend 'docker' using platform 'linux/amd64' running up to 1 pipelines in parallel"}
Feb 24 08:21:59 yorgos woodpecker-agent[277462]: {"level":"error","error":"Error response from daemon: no container with name or ID \"wp_01JMVGRX6G3F4258199GHYCB14\" found: no such container","time":"2025-02-24T08:21:59Z","message":"could not kill container 'clone'"}
Feb 24 08:21:59 yorgos woodpecker-agent[277462]: {"level":"error","error":"Error response from daemon: no container with ID or name \"wp_01JMVGRX6G3F4258199GHYCB14\" found: no such container","time":"2025-02-24T08:21:59Z","message":"could not remove container 'clone'"}
Feb 24 08:21:59 yorgos woodpecker-agent[277462]: {"level":"error","error":"Error response from daemon: no container with name or ID \"wp_01JMVGRX6G3F4258199HG5VQW6\" found: no such container","time":"2025-02-24T08:21:59Z","message":"could not kill container 'Lint'"}
Feb 24 08:21:59 yorgos woodpecker-agent[277462]: {"level":"error","error":"Error response from daemon: no container with ID or name \"wp_01JMVGRX6G3F4258199HG5VQW6\" found: no such container","time":"2025-02-24T08:21:59Z","message":"could not remove container 'Lint'"}
Feb 24 08:21:59 yorgos woodpecker-agent[277462]: {"level":"error","error":"rpc error: code = Unknown desc = workflow finished with error uuid=01JMVGRX6G3F4258199GHYCB14: exit code 128","time":"2025-02-24T08:21:59Z","message":"grpc error: wait(): code: Unknown"}
Feb 24 08:21:59 yorgos woodpecker-agent[277462]: {"level":"warn","repo":"redacted/myproject","pipeline":"6","workflow_id":"19","error":"rpc error: code = Unknown desc = workflow finished with error uuid=01JMVGRX6G3F4258199GHYCB14: exit code 128","time":"2025-02-24T08:21:59Z","message":"cancel signal received"}
```

It does look like a problem the agent has in spawning the containers...

**Comment by pat-s** [comment]
It is only a question of time before this continues to break (again) as it is using the docker/moby libs in the background instead of the (existing) native podman ones.

I would advise against using podman like this and rather encourage to put the time in for an official podman backend.

**Comment by gsaslis** [comment]
Thanks for clarifying @pat-s ! Appreciate the words of caution and it's something I'll definitely keep in mind.

**Comment by Darin755** [comment]
I'm using rootful Podman but the pipeline fails at the clone step. I have Forgejo, Woodpecker and Woodpecker agent all in the same pod and when I run the pipeline I get the following error:

```
fatal: unable to access http://localhost:3001/darin755/woodpeckerhello.git: failed to connect to localhost port 3001 after 1 ms: Could not connect to server
```

Is this a known problem or should I create a new issue?

**Comment by anbraten** [comment]
> Is this a known problem or should I create a new issue?

That's unrelated to podman and seems to be rather a configuration issue with forgejo as forgejo is saying it's public address is localhost:... which is not the case for a container running the pipeline.

**Comment by Darin755** [comment]
> > Is this a known problem or should I create a new issue?
> 
> That's unrelated to podman and seems to be rather a configuration issue with forgejo as forgejo is saying it's public address is localhost:... which is not the case for a container running the pipeline.

That's what I set it to for testing purposes. I'm just running it locally to play around with it. I might properly deploy it in the future but for now I just want to learn the basics on my local machine.

**Comment by duckinator** [comment]
On Debian 13 (trixie), I did the following:

1. Run `systemctl --user enable --now podman.socket` which creates `/run/user/$UID/podman/podman.sock` (in my case, UID=1000).
2. Add this to my Bash config: `export WOODPECKER_BACKEND_DOCKER_HOST="unix:///run/user/$UID/podman/podman.sock"`

And now `woodpecker-cli exec` uses podman via the docker backend, without issue.

A native backend would be nice just because it wouldn't need the extra config, but this seems like a pretty good workaround until that happens.
